lib/common/JSONImporter/NodeStatePatch.ts

```diff
@@ -0,0 +1,519 @@
+import {NodeSelections} from './NodeSelectors';
+import {NodeChangeSet} from './NodeChangeSet';
+import {NodeSearchUtils, setNested} from './Utils';
+import {Maybe, Result} from 'ts-monads';
+import diff, {ChangeSet, ChangeType} from 'changeset';
+import JSONImporter from "../JSONImporter";
+import Core = GmeClasses.Core;
+
+type PatchFunction = (node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections) => Promise<Result<PatchResult, PatchError>>;
+type PatchResultPromise = Promise<Result<PatchResult, PatchError>>;
+
+class PatchResult {
+    patches: NodeChangeSet | NodeChangeSet[];
+
+    constructor(patches: NodeChangeSet | NodeChangeSet[]) {
+        this.patches = patches;
+    }
+}
+
+export class PatchError extends Error {
+    constructor(msg: string) {
+        super(msg);
+    }
+}
+
+export interface PatchOperation {
+    core: GmeClasses.Core;
+    put: PatchFunction;
+    delete: PatchFunction;
+}
+
+export abstract class NodeStatePatch implements PatchOperation {
+    core: GmeClasses.Core;
+    nodeSearchUtils: NodeSearchUtils
+
+    constructor(core: GmeClasses.Core, nodeSearchUtils: NodeSearchUtils) {
+        this.core = core;
+        this.nodeSearchUtils = nodeSearchUtils;
+    }
+
+    keyLengthValidator(change) {
+        return change.key.length === 2;
+    }
+
+    async put(node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections): PatchResultPromise {
+        return this._validChange(change)
+            .map(validChange => this._put(node, change, resolvedSelectors));
+    }
+
+    async delete(node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections): PatchResultPromise {
+        return this._validChange(change)
+            .map(validChange => this._delete(node, change, resolvedSelectors));
+    }
+
+    _validChange(change: NodeChangeSet): Result<NodeChangeSet, PatchError> {
+        return Result.Ok(change);  // override this as needed in child classes
+    }
+
+    abstract _delete: PatchFunction;
+    abstract _put: PatchFunction;
+}
+
+export class AttributesPatch extends NodeStatePatch {
+    _validChange(change: NodeChangeSet): Result<NodeChangeSet, PatchError> {
+        if (change.key.length === 2) {
+            return Result.Ok(change);
+        }
+        const msg = `Complex attributes not currently supported: ${change.key.join(', ')}`;
+        return Result.Error(PatchError(msg));
+    }
+
+    _put = async (node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections): PatchResultPromise => {
+        const [/*type*/, name] = change.key;
+        this.core.setAttribute(node, name, change.value || '');
+        return new PatchResult(change);  // I am not sure we need this PatchResult as it seems that it only ever returns the NodeChangeSet that was passed to it...?
```

Comment:
I didn't remove `PatchResult` but does this ever return something other than the change that was applied?

Reply:
Based on your suggestion, I have created a `PatchInfo` object and the signature of `_put` and `_delete` methods go as follows:
```ts
class PatchInfo {
    nodeId: GmeCommon.Path;
    nodeGuid: Core.GUID;
    target: string;
    appliedPatch: ChangeSet;
}

type PatchFunction = (node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections) => Promise<void>;

```